llvm/include/llvm/CAS/CASFileSystem.h

```diff
@@ -9,14 +9,59 @@
 #ifndef LLVM_CAS_CASFILESYSTEM_H
 #define LLVM_CAS_CASFILESYSTEM_H
 
-#include <llvm/CAS/ThreadSafeFileSystem.h>
+#include "llvm/Support/ExtensibleRTTI.h"
 #include <llvm/Support/Error.h>
+#include <llvm/Support/VirtualFileSystem.h>
 
-namespace llvm {
-namespace cas {
+namespace llvm::cas {
 class ObjectStore;
 class CASID;
 
+/// Abstract class represents an open file backed by a CAS.
+class CASBackedFile : public RTTIExtends<CASBackedFile, vfs::File> {
+public:
+  static const char ID;
+  /// Get the CAS reference for the contents of the file.
+  virtual cas::ObjectRef getObjectRefForContent() = 0;
+};
+
+/// Abstract class represents a CAS backed file system.
+class CASBackedFileSystem
+    : public llvm::RTTIExtends<CASBackedFileSystem, vfs::FileSystem> {
+public:
+  static const char ID;
+
+  /// This is a convenience method that opens a file, gets its content and then
+  /// closes the file. It returns MemoryBuffer and ObjectRef in one call to avoid
+  /// open the file twice.
+  /// The IsText parameter is used to distinguish whether the file should be
+  /// opened as a binary or text file.
+  llvm::Expected<std::pair<std::unique_ptr<llvm::MemoryBuffer>, cas::ObjectRef>>
+  getBufferAndObjectRefForFile(const Twine &Name, int64_t FileSize = -1,
+                               bool RequiresNullTerminator = true,
+                               bool IsVolatile = false, bool IsText = true);
+
+  /// Get ObjectRef of a file from its path.
+  llvm::Expected<cas::ObjectRef> getObjectRefForFileContent(const Twine &Name);
+
+  /// Implementation for openFileForRead using CASBackedFile.
+  ErrorOr<std::unique_ptr<vfs::File>>
+  openFileForRead(const Twine &Path) override {
+    auto F = openCASBackedFileForRead(Path);
+    if (!F)
+      return errorToErrorCode(F.takeError());
+    return std::move(*F);
+  }
+
+  /// Get CASBackedFile for read.
+  virtual llvm::Expected<std::unique_ptr<CASBackedFile>>
+  openCASBackedFileForRead(const Twine &Path) = 0;
```

Comment:
What's the reason for having this be a separate API instead of just using `openFileForRead` and `cast<>`ing the result to a `CASBackedFile`?  Having both be virtual allows the posibility that a subclass returns different files from each one.